packer_arm_substitute.py

```diff
@@ -0,0 +1,285 @@
+#!/usr/bin/env python3
+
+# Python script that roughly mimics the workings of packer on ARM,
+# where it is not well supported
+#
+# 1. Given a vars.json file, download the iso specified by those vars to a cache
+#    location
+# 2. Start an automated VM install using that iso as a base, passing in a 
+#    kickstart file via extra kernel args using libvirt
+# 3. Transfer a set of config files from the host into the guest via a volume mount
+# 4. Attempt to log into the guest, copy the files into the local image, and run several shell commands
+# 5. Optionally, transfer the resultant VM disk image from the libvirt storage pool to a location on disk
+
+import pexpect
+import json
+import requests
+import hashlib
+from pathlib import Path
+import subprocess
+from time import sleep
+import argparse
+import random
+from datetime import datetime, timedelta
+
+# Attempt to cache things in a manner similar to packer
+
+CACHE_DIR = Path() / 'packer_cache'
+CHUNK_SIZE = 2**16
+
+def hash_iso(iso_path: Path, algorithm: str = 'sha256') -> str:
+    hash_func = hashlib.new(algorithm)
+    with open(iso_path, 'rb') as iso_f:
+        while chunk := iso_f.read(CHUNK_SIZE):
+            hash_func.update(chunk)
+    
+    return hash_func.hexdigest()
+
+def download_iso(iso_dir: str, vars_path: Path) -> Path:
+    ''' Given a vars.json, '''
+    with open(vars_path, 'r') as varsf:
+        iso_vars = json.loads(varsf.read())
+
+    # TODO packer has a naming scheme for downloaded ISOs that's hard to replicate
```

Comment:
Apparently, it's the sha1sum of the value of iso_checksum in the vars.json file (plus a .iso extension).